turbot/resource_turbot_file.go

```diff
@@ -0,0 +1,238 @@
+	id := d.Id()
+	// pass nil
+	resource, err := client.ReadFullResource(id)
+	if err != nil {
+		if apiClient.NotFoundError(err) {
+			// resource was not found - clear id
+			d.SetId("")
+		}
+		return err
+	}
+
+	// rebuild data from the resource
+	content, err := helpers.MapToJsonString(resource.Data)
+	if err != nil {
+		return fmt.Errorf("error building resource content: %s", err.Error())
+	}
+
+	customMetadata := resource.Turbot.Custom
+
+	// set parent_akas property by loading resource and fetching the akas
+	if err := storeAkas(resource.Turbot.ParentId, "parent_akas", d, meta); err != nil {
+		return err
+	}
+
+	if v, ok := customMetadata["title"]; ok {
+		d.Set("title", v)
+	}
+	if v, ok := customMetadata["description"]; ok {
+		d.Set("description", v)
+	}
+	// assign results back into ResourceData
+	d.Set("parent", resource.Turbot.ParentId)
+	d.Set("content", content)
+	return nil
+}
+
+func resourceTurbotFileUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	// build input map to pass to mutation
+	id := d.Id()
+
+	input, err := buildFileInput(d, fileProperties)
+	if err != nil {
+		return err
+	}
+	input["id"] = id
+
+	turbotMetadata, err := client.UpdateResource(input)
+	if err != nil {
+		return err
+	}
+	// save the formatted data: this is to ensure the acceptance tests behave in a consistent way regardless of the ordering of the json data
+	d.Set("content", helpers.FormatJson(d.Get("content").(string)))
+
+	metadataMap := turbotMetadata.Custom
+	if v, ok := metadataMap["description"]; ok {
+		d.Set("description", v)
+	}
+	d.Set("title", metadataMap["title"])
+	// set parent_akas property by loading resource and fetching the akas
+	return storeAkas(turbotMetadata.ParentId, "parent_akas", d, meta)
+}
+
+func resourceTurbotFileDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	id := d.Id()
+	err := client.DeleteResource(id)
+	if err != nil {
+		return err
+	}
+
+	// clear the id to show we have deleted
+	d.SetId("")
+
+	return nil
+}
+
+func resourceTurbotFileImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	if err := resourceTurbotResourceRead(d, meta); err != nil {
+		return nil, err
+	}
+	return []*schema.ResourceData{d}, nil
+}
+
+func buildFileInput(d *schema.ResourceData, properties []interface{}) (map[string]interface{}, error) {
+	var err error
+	var input = make(map[string]interface{})
+	var content = make(map[string]interface{})
+	input = mapFromResourceData(d, properties)
+	// convert data from json string to map
+	contentString := d.Get("content").(string)
+	if content, err = helpers.JsonStringToMap(contentString); err != nil {
+		return nil, fmt.Errorf("error build resource mutation input, failed to unmarshal content: \n%s\nerror: %s", contentString, err.Error())
+	}
+	// set old properties of content to NULL in update operation
+	if old, new := d.GetChange("content"); old != nil {
```

Comment:
this should be split out from buildFileInput - we only want to do this for an update